[SimplifyCFG] Consider preds to switch in `simplifyDuplicateSwitchArms`

antoniofrighetto opened the pull request:
Allow a duplicate basic block with multiple predecessors to the jump table to be simplified, by considering that the same basic block may appear in more switch cases.

dtcxzyw commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -7575,8 +7573,10 @@ bool SimplifyCFGOpt::simplifyDuplicateSwitchArms(SwitchInst *SI,
-    // SI need to be updated.
-    if (BB->hasNPredecessorsOrMore(2))
+    // SI need to be updated. For now, consider only backedges to the SI.
+    if (pred_size(BB) > 4 || any_of(predecessors(BB), [&](BasicBlock *Pred) {
```
Use `hasNPredecessorsOrMore(4)` and `BB->getUniquePredecessor() == SI->getParent()`.

michaelmaitland replied:
What is special about having >= 4 predecessors?

dtcxzyw replied:
It is used to avoid `O(N^2)` lookup. We may choose a larger threshold if the compile-time impact is negligible.

dtcxzyw reviewed (approved):
LGTM. Please wait for additional approval from other reviewers.

michaelmaitland commented on llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table.ll:
```diff
@@ -340,15 +340,23 @@ entry:
-sw.bb3: br label %if.end
-sw.default: br label %if.end
-if.else: br label %if.end
+sw.bb:                                            ; preds = %entry, %entry
```
Did you mean to change these lines?

antoniofrighetto replied:
I took the chance to regenerate the test too, as the `if.else` branch was dead.

nikic reviewed (commented):
Can you please pre-commit the tests and add a test where the transform triggers but does not convert into a select afterwards (only reduces blocks)?

michaelmaitland reviewed (approved):
LGTM

antoniofrighetto committed:
[SimplifyCFG] Precommit tests for PR118955 (NFC)

antoniofrighetto committed:
[SimplifyCFG] Consider preds to switch in `simplifyDuplicateSwitchArms`

Allow a duplicate basic block with multiple predecessors to the
jump table to be simplified, by considering that the same basic
block may appear in more switch cases.
